@This does not work with @RequiresCalledMethods when chaining

I'm observing a behavior where it seems that `@This` (i.e. the Returns Receiver Checker) does not properly forward the `@CalledMethods` qualifier of the object when chaining the result to a method that uses `@RequiresCalledMethods` to define requirements for the receiver.

Chaining the method call works when the requirements are defined with `@CalledMethods` on the explicit receiver parameter:

```java
import org.checkerframework.checker.calledmethods.qual.CalledMethods;
import org.checkerframework.checker.calledmethods.qual.RequiresCalledMethods;
import org.checkerframework.common.returnsreceiver.qual.This;

public class Foo {

    public @This Foo bar() {
        return this;
    }

    public void baz(@CalledMethods("bar") Foo this) {}

    public static void test() {
        final Foo foo = new Foo();
        foo.bar().baz(); // No error
    }

}
```

Storing the intermediate result, then calling a `@RequiresCalledMethods`-annotated method also works:

```java
import org.checkerframework.checker.calledmethods.qual.RequiresCalledMethods;
import org.checkerframework.common.returnsreceiver.qual.This;

public class Foo {

    public @This Foo bar() {
        return this;
    }

    @RequiresCalledMethods( value = "this", methods = "bar" )
    public void baz() {}

    public static void test() {
        final Foo foo = new Foo();
        final Foo foo1 = foo.bar();
        foo1.baz(); // No error
    }

}
```

But if I chain the call to the `@RequiresCalledMethods`-annotated method:

```java
import org.checkerframework.checker.calledmethods.qual.RequiresCalledMethods;
import org.checkerframework.common.returnsreceiver.qual.This;

public class Foo {

    public @This Foo bar() {
        return this;
    }

    @RequiresCalledMethods( value = "this", methods = "bar" )
    public void baz() {}

    public static void test() {
        final Foo foo = new Foo();
        foo.bar().baz(); // Error
    }

}
```

I get the error

```
> .\bin\javac.bat -cp . -processor calledmethods Foo.java
Foo.java:42: error: [contracts.precondition] precondition of foo.bar().baz is not satisfied.
        foo.bar().baz();
                     ^
  found   : foo.bar() is @CalledMethods
  required: foo.bar() is @CalledMethods("bar")
1 error
```

Which indicates that the qualifier isn't getting forwarded.

While this can be worked around by explicitly defining the `this` parameter and using `@CalledMethods`, it can become quite verbose when the class name is long and/or there are type parameters involved (imagine having to add a `@CalledMethods({"foo", "bar", "baz"}) VeryLongClassName<E, T, I> this` parameter on multiple methods in a class), and (IMO) is much less clear at a glance than a `@RequiresCalledMethods` annotation.

All of the examples were ran using the command `.\bin\javac.bat -cp . -processor calledmethods Foo.java`, with the code being placed in the `Foo.java` file, and with the current directory being the root directory of the Checker 3.27.0 release artifact.

@msridhar I would have thought so, too, but I just tried it out and it does not prevent the error from being issued. That might be a (separate) bug with that command-line option. In particular, I followed the instructions in the original issue report. I used the following `Foo.java`:
```
import org.checkerframework.common.returnsreceiver.qual.This;

public class Foo {

    public @This Foo bar() {
        return this;
    }

    @RequiresCalledMethods( value = "this", methods = "bar" )
    public void baz() {}

    public static void test() {
        final Foo foo = new Foo();
        foo.bar().baz(); // Error
    }

}
```

I ran this command:
```
./checker/bin/javac -cp . -processor calledmethods -AassumeDeterministic Foo.java
```

And got this error (i.e., the same one reported originally):
```
Foo.java:15: error: [contracts.precondition] precondition of foo.bar().baz is not satisfied.
        foo.bar().baz(); // Error
                     ^
  found   : foo.bar() is @CalledMethods
  required: foo.bar() is @CalledMethods("bar")
1 error
```

If I add an `@Deterministic` annotation (as in my previous comment) to `Foo.java`, I don't get an error from the typechecker as expected. @mernst, is this a bug with `-AassumeDeterministic`?

Interesting, that's good to know. Using `@Deterministic` on the methods is not too bad (I was already doing that anyway), though it sure would be convenient if there was a way to automatically make a method `@Deterministic` if it has a `@This` return since it is always true by definition (but that's OK). I agree that it definitely should be documented somewhere though.

However, there's still a problem. If the method that needs to be called receives a parameter, and that parameter is an expression instead of a constant or variable, it still won't work, even with `@Deterministic`:

```java
import org.checkerframework.checker.calledmethods.qual.RequiresCalledMethods;
import org.checkerframework.common.returnsreceiver.qual.This;
import org.checkerframework.dataflow.qual.Deterministic;

public class Foo {

    @Deterministic
    public @This Foo bar(String s) {
        return this;
    }

    @RequiresCalledMethods( value = "this", methods = "bar" )
    public void baz() {}

    public static void test() {
        final Foo foo1 = new Foo();
        foo1.bar("").baz(); // This is fine
        var s = String.format("");
        final Foo foo2 = new Foo();
        foo2.bar(s).baz(); // This is also fine
        final Foo foo3 = new Foo();
        foo3.bar(String.format("")).baz(); // This is not
    }

}
```
```
.\bin\javac.bat -cp . -processor calledmethods Foo.java
Foo.java:55: error: [contracts.precondition] precondition of foo3.bar(String.format("")).baz is not satisfied.
        foo3.bar(String.format("")).baz(); // This is not
                                       ^
  found   : foo3.bar(java.lang.String.format("")) is @CalledMethods
  required: foo3.bar(java.lang.String.format("")) is @CalledMethods("bar")
1 error
```

Using an anonymous object for the original receiver also causes issues:

```java
import org.checkerframework.checker.calledmethods.qual.RequiresCalledMethods;
import org.checkerframework.common.returnsreceiver.qual.This;
import org.checkerframework.dataflow.qual.Deterministic;

public class Foo {

    @Deterministic
    public @This Foo bar() {
        return this;
    }

    @RequiresCalledMethods( value = "this", methods = "bar" )
    public void baz() {}

    public static void test() {
        final Foo foo = new Foo();
        foo.bar().baz(); // Fine
        new Foo().bar().baz(); // Not fine
    }

}
```
```
.\bin\javac.bat -cp . -processor calledmethods Foo.java
Foo.java:45: error: [contracts.precondition] precondition of new Foo().bar().baz is not satisfied.
        new Foo().bar().baz(); // Not fine
                           ^
  found   : new Foo().bar() is @CalledMethods
  required: new Foo().bar() is @CalledMethods("bar")
1 error
```

Given the context I would imagine it to be some sort of limitation of how `@Deterministic` is applied. Like before, if it's not fixable it should definitely be documented, though it really is unfortunate that it makes `@RequiresCalledMethods` a lot less useful.

--- a/checker/tests/calledmethods/Issue5402.java
+++ b/checker/tests/calledmethods/Issue5402.java
@@ -1,26 +1,70 @@
+// Test case for https://github.com/typetools/checker-framework/issues/5402
+
+import org.checkerframework.checker.calledmethods.qual.CalledMethods;
 import org.checkerframework.checker.calledmethods.qual.RequiresCalledMethods;
 import org.checkerframework.common.returnsreceiver.qual.This;
-import org.checkerframework.checker.calledmethods.qual.CalledMethods;
+import org.checkerframework.dataflow.qual.Deterministic;
 
-public class Issue5402 {
+public class Issue5402 {}
 
-  public @This Issue5402 bar() {
+class Issue5402_Ok1 {
+
+  public @This Issue5402_Ok1 bar() {
     return this;
   }
-  public @This Issue5402 bar(String s) {
-    return this;
-  }
 
-  @RequiresCalledMethods( value = "this", methods = "bar" )
-  public void baz() {}
-
+  public void baz(@CalledMethods("bar") Issue5402_Ok1 this) {}
 
   public static void test() {
-    final Issue5402 foo = new Issue5402();
-    foo.bar().baz();
-    new Issue5402().bar().baz();
-     final Issue5402 foo3 = new Issue5402();
-        foo3.bar(String.format("")).baz();
+    final Issue5402_Ok1 foo = new Issue5402_Ok1();
+    foo.bar().baz(); // No error
+  }
+}
+
+class Issue5402_Ok2 {
+
+  public @This Issue5402_Ok2 bar() {
+    return this;
   }
 
+  @RequiresCalledMethods(value = "this", methods = "bar")
+  public void baz() {}
+
+  public static void test() {
+    final Issue5402_Ok2 foo = new Issue5402_Ok2();
+    final Issue5402_Ok2 foo1 = foo.bar();
+    foo1.baz(); // No error
+  }
+}
+
+class Issue5402_Ok3 {
+
+  @Deterministic
+  public @This Issue5402_Ok3 bar() {
+    return this;
+  }
+
+  @RequiresCalledMethods(value = "this", methods = "bar")
+  public void baz() {}
+
+  public static void test() {
+    final Issue5402_Ok2 foo = new Issue5402_Ok2();
+    final Issue5402_Ok2 foo1 = foo.bar();
+    foo1.baz(); // No error
+  }
+}
+
+class Issue5402_Bad {
+
+  public @This Issue5402_Bad bar() {
+    return this;
+  }
+
+  @RequiresCalledMethods(value = "this", methods = "bar")
+  public void baz() {}
+
+  public static void test() {
+    final Issue5402_Bad foo = new Issue5402_Bad();
+    foo.bar().baz(); // Error
+  }
 }

--- a/checker/tests/calledmethods/Issue5402.java
+++ b/checker/tests/calledmethods/Issue5402.java
@@ -1,71 +1,29 @@
-// Test case for https://github.com/typetools/checker-framework/issues/5402
-// @skip-test until the bug is fixed
-
-import org.checkerframework.checker.calledmethods.qual.CalledMethods;
 import org.checkerframework.checker.calledmethods.qual.RequiresCalledMethods;
 import org.checkerframework.common.returnsreceiver.qual.This;
+import org.checkerframework.checker.calledmethods.qual.CalledMethods;
 import org.checkerframework.dataflow.qual.Deterministic;
 
-public class Issue5402 {}
-
-class Issue5402_Ok1 {
-
-  public @This Issue5402_Ok1 bar() {
-    return this;
-  }
-
-  public void baz(@CalledMethods("bar") Issue5402_Ok1 this) {}
-
-  public static void test() {
-    final Issue5402_Ok1 foo = new Issue5402_Ok1();
-    foo.bar().baz(); // No error
-  }
-}
-
-class Issue5402_Ok2 {
-
-  public @This Issue5402_Ok2 bar() {
-    return this;
-  }
-
-  @RequiresCalledMethods(value = "this", methods = "bar")
-  public void baz() {}
-
-  public static void test() {
-    final Issue5402_Ok2 foo = new Issue5402_Ok2();
-    final Issue5402_Ok2 foo1 = foo.bar();
-    foo1.baz(); // No error
-  }
-}
-
-class Issue5402_Ok3 {
+public class Issue5402 {
 
   @Deterministic
-  public @This Issue5402_Ok2 bar() {
+  public @This Issue5402 bar() {
+    return this;
+  }
+  @Deterministic
+  public @This Issue5402 bar(String s) {
     return this;
   }
 
-  @RequiresCalledMethods(value = "this", methods = "bar")
+  @RequiresCalledMethods( value = "this", methods = "bar" )
   public void baz() {}
 
-  public static void test() {
-    final Issue5402_Ok2 foo = new Issue5402_Ok2();
-    final Issue5402_Ok2 foo1 = foo.bar();
-    foo1.baz(); // No error
-  }
-}
-
-class Issue5402_Bad {
-
-  public @This Issue5402_Bad bar() {
-    return this;
-  }
-
-  @RequiresCalledMethods(value = "this", methods = "bar")
-  public void baz() {}
 
   public static void test() {
-    final Issue5402_Bad foo = new Issue5402_Bad();
-    foo.bar().baz(); // Error
+    final Issue5402 foo = new Issue5402();
+    foo.bar().baz();
+    new Issue5402().bar().baz();
+     final Issue5402 foo3 = new Issue5402();
+        foo3.bar(String.format("")).baz();
   }
+
 }

--- a/checker/tests/calledmethods/Issue5402.java
+++ b/checker/tests/calledmethods/Issue5402.java
@@ -1,15 +1,12 @@
 import org.checkerframework.checker.calledmethods.qual.RequiresCalledMethods;
 import org.checkerframework.common.returnsreceiver.qual.This;
 import org.checkerframework.checker.calledmethods.qual.CalledMethods;
-import org.checkerframework.dataflow.qual.Deterministic;
 
 public class Issue5402 {
 
-  @Deterministic
   public @This Issue5402 bar() {
     return this;
   }
-  @Deterministic
   public @This Issue5402 bar(String s) {
     return this;
   }

--- a/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java
+++ b/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java
@@ -1835,18 +1835,29 @@ public class BaseTypeVisitor<Factory extends GenericAnnotatedTypeFactory<?, ?, ?
         Set<AnnotationMirror> annos = value.getAnnotations();
         inferredAnno = hierarchy.findAnnotationInSameHierarchy(annos, anno);
       } else {
-        // If there is no information in the store (possible if e.g., no refinement
-        // of the field has occurred), use top instead of automatically
-        // issuing a warning. This is not perfectly precise: for example,
-        // if jeExpr is a field it would be more precise to use the field's
-        // declared type rather than top. However, doing so would be unsound
-        // in at least three circumstances where the type of the field depends
-        // on the type of the receiver: (1) all fields in Nullness Checker,
-        // because of possibility that the receiver is under initialization,
-        // (2) polymorphic fields, and (3) fields whose type is a type variable.
-        // Using top here instead means that there is no need for special cases
-        // for these situations.
-        inferredAnno = atypeFactory.getQualifierHierarchy().getTopAnnotation(anno);
+        if (expressionString.equals("this")) {
+          AnnotatedTypeMirror atype = atypeFactory.getReceiverType(tree);
+          if (atype != null) {
+            QualifierHierarchy hierarchy = atypeFactory.getQualifierHierarchy();
+            Set<AnnotationMirror> annos = atype.getEffectiveAnnotations();
+            inferredAnno = hierarchy.findAnnotationInSameHierarchy(annos, anno);
+          }
+        }
+
+        if (inferredAnno == null) {
+          // If there is no information in the store (possible if e.g., no refinement
+          // of the field has occurred), use top instead of automatically
+          // issuing a warning. This is not perfectly precise: for example,
+          // if jeExpr is a field it would be more precise to use the field's
+          // declared type rather than top. However, doing so would be unsound
+          // in at least three circumstances where the type of the field depends
+          // on the type of the receiver: (1) all fields in Nullness Checker,
+          // because of possibility that the receiver is under initialization,
+          // (2) polymorphic fields, and (3) fields whose type is a type variable.
+          // Using top here instead means that there is no need for special cases
+          // for these situations.
+          inferredAnno = atypeFactory.getQualifierHierarchy().getTopAnnotation(anno);
+        }
       }
       if (!checkContract(exprJe, anno, inferredAnno, store)) {
         if (exprJe != null) {

--- a/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java
+++ b/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java
@@ -1835,6 +1835,8 @@ public class BaseTypeVisitor<Factory extends GenericAnnotatedTypeFactory<?, ?, ?
         Set<AnnotationMirror> annos = value.getAnnotations();
         inferredAnno = hierarchy.findAnnotationInSameHierarchy(annos, anno);
       } else {
+        // If the expression is "this", then get the type of the method receiver.
+        // TODO: There are other expressions that can be converted to trees, "#1" for example.
         if (expressionString.equals("this")) {
           AnnotatedTypeMirror atype = atypeFactory.getReceiverType(tree);
           if (atype != null) {

